File: Golang/RecursionExtra/totalSum.go
/*

Bài tập đệ quy: Tính tổng chữ số của một số nguyên dương

Đề bài:
Viết hàm đệ quy sumDigits(n) trả về tổng các chữ số của số nguyên dương n.

Ví dụ:

sumDigits(1234) = 1 + 2 + 3 + 4 = 10

sumDigits(505) = 5 + 0 + 5 = 10

*/

// this is the way that I use my brain to think but it kinda sucks, because recursion usually avoid accumulator
package main

func sumDigits(number int, counter int) int {
	// base case
	if number/10 == 0 {
		return number%10 + counter
	} else {
		counter += number % 10
		return sumDigits(number/10, counter)
	}

}

func sumDigitsBestWay(number int) int {
	if number < 10 {
		return number
	}

	return sumDigitsBestWay(number/10) + (number % 10)
}

File: TypeScript/RecursionExtra/totalSum.ts
/*

Bài tập đệ quy: Tính tổng chữ số của một số nguyên dương

Đề bài:
Viết hàm đệ quy sumDigits(n) trả về tổng các chữ số của số nguyên dương n.

Ví dụ:

sumDigits(1234) = 1 + 2 + 3 + 4 = 10

sumDigits(505) = 5 + 0 + 5 = 10

*/

const sumDigits = (n: number): number => {
    if (n === 0) {
        return 0;
    } else {
        return (n % 10) + sumDigits(Math.floor(n / 10));
    }
};

// Test cases
console.log(sumDigits(1234)); // Output: 10
console.log(sumDigits(505));  // Output: 10
console.log(sumDigits(0));    // Output: 0
console.log(sumDigits(9));    // Output: 9
console.log(sumDigits(1001)); // Output: 2